Add a Latest Comments block

danielbachhuber opened the pull request:
* [x] Permit number of comments displayed to be configurable.
* [x] Permit comment excerpt to be shown.
* [x] Permit avatar to be shown.
* [x] Permit timestamps to be shown.
* [x] Register server-side block for rendering comments on frontend.
* [ ] Handle truncated comments as excerpts in editor.

See #1792

danielbachhuber committed:
First pass at a basic Latest Comments block

danielbachhuber committed:
Add fixtures for latest comments block

danielbachhuber committed:
Add controls for toggling avatar, excerpt and timestamp

danielbachhuber committed:
Render timestamp when `displayTimestamp=true`

danielbachhuber committed:
Use the actual post title in the comment summary

danielbachhuber committed:
Render avatar when `displayAvatar=true`

danielbachhuber committed:
Render excerpt when `displayExcerpt=true`

danielbachhuber committed:
Put controls in display order

danielbachhuber commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,209 @@
+			super( ...arguments );
+			this.toggleDisplayAvatar = this.toggleDisplayAvatar.bind( this );
+			this.toggleDisplayExcerpt = this.toggleDisplayExcerpt.bind( this );
+			this.toggleDisplayTimestamp = this.toggleDisplayTimestamp.bind( this );
```
It would be nice if there was some common abstraction for toggling boolean attributes.

westonruter replied:
How about this? https://github.com/WordPress/gutenberg/pull/1931/commits/275b7d9303fb07e333a61f793e0e91f9d1835dfa

danielbachhuber replied:
👌

danielbachhuber commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,209 @@
+						/>
+
+						<ToggleControl
+							label={ __( 'Display timestamp' ) }
```
Should this be "Display timestamp" or "Display date"? It'd be nice to have language consistent with posts.

westonruter replied:
I think it should be “Display publish date” both here and on the Latest Posts block.

westonruter replied:
Or yeah, just “Display date” would be fine I think.

danielbachhuber replied:
> I think it should be “Display publish date” both here and on the Latest Posts block

Should I change this in a follow-up PR?

danielbachhuber commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,209 @@
+				<ul className={ this.props.className } key="latest-comments">
+					{ latestComments.map( ( comment, i ) =>
+						<li key={ i }>
+							{ displayAvatar && comment.author_avatar_urls[ 96 ] &&
```
Is this `96` a safe hard-coded assumption? Or, should we use the largest avatar size on the object?

westonruter replied:
No, it's not safe because `rest_avatar_sizes` can be filtered to cause different sizes to be returned in the response. So yeah, using the largest size I think is the right move.

danielbachhuber commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,209 @@
+							{ displayAvatar && comment.author_avatar_urls[ 96 ] &&
+								<img className={ `${ this.props.className }__comment-avatar` } alt={ comment.author_name } src={ comment.author_avatar_urls[ 96 ] } />
+							}
+							<a href={ comment.link } target="_blank">{ comment._embedded.up[ 0 ].title.rendered }</a>
```
What should we display if there's no linked post, or the linked post doesn't have a title?

westonruter replied:
When there is no title, see 813af2150f4c98d55139a4533711e54be44d2dc6.

danielbachhuber replied:
Updated in f3aa112f

danielbachhuber commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,209 @@
+							<a href={ comment.link } target="_blank">{ comment._embedded.up[ 0 ].title.rendered }</a>
+							{ displayTimestamp && comment.date_gmt &&
+								<span className={ `${ this.props.className }__comment-timestamp` }>
+									{ moment( comment.date_gmt ).local().format( 'MMM DD h:mm A' ) }
```
This timestamp format seems like it should be user-configurable. Maybe we need a JS implementation of `human_time_diff()` with a filterable format.

westonruter replied:
Moment.js does support relative time: https://momentjs.com/docs/#/displaying/fromnow/

It could be a nice block attribute to decide whether the dates are shown in relative time or absolute time, according to the site's datetime format. 

See also https://github.com/WordPress/gutenberg/pull/1992#issuecomment-317650041 and https://github.com/WordPress/gutenberg/pull/1992#issuecomment-317657482

danielbachhuber replied:
> according to the site's datetime format.

Should we get a new issue going about respecting the site's datetime format?

westonruter replied:
@danielbachhuber done: https://github.com/WordPress/gutenberg/issues/2013

mtias replied:
Yes, it affects latest posts as well.

danielbachhuber commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,209 @@
+								</span>
+							}
+							{ displayExcerpt && comment.content &&
+								<div className={ `${ this.props.className }__comment-excerpt` } dangerouslySetInnerHTML={ { __html: comment.content.rendered } } />
```
What algorithm would you like to use for truncating comments? WordPress doesn't currently implement a comment excerpt.

westonruter replied:
Why the `dangerouslySetInnerHTML` here? Because a user may have markup in their excerpt?

I am not aware of an existing truncation solution, other than to use CSS.

danielbachhuber replied:
> Why the `dangerouslySetInnerHTML` here? Because a user may have markup in their excerpt?

Yes, HTML is permitted in comments (and WordPress includes some by default anyway)

westonruter replied:
@mtias thoughts on whether raw HTML from comments should be embedded?

danielbachhuber commented on blocks/library/latest-comments/style.scss:
```diff
@@ -0,0 +1,23 @@
+		height: 48px;
+		width: 48px;
+		border-radius: 24px;
+		margin-right: 5px;
```
Is it fine that these sizes are in px? If not, can you give me the requested values as em?

westonruter replied:
There are 30 instances of `em` units being used in block CSS, whereas there are 37 instances of `px` units being used. So I'd say `px` is safe. 👍

danielbachhuber commented on blocks/test/fixtures/core__latest-comments.json:
```diff
@@ -0,0 +1,12 @@
+            "displayTimestamp": true
+        }
+    }
+]
```
It'd be nice to set up mock data so we can actually test component rendering.

danielbachhuber reviewed (commented):
@westonruter @melchoyce This is ready for an initial review. Here's how it looks:

![image](https://user-images.githubusercontent.com/36432/28576656-48b0716e-7109-11e7-8515-741fd2856e75.png)

melchoyce commented:
Very cool — works as expected to me. Once the your code points are reviewed, let's commit and I'll see if I can figure out a follow-up PR to tweak the design/CSS.

danielbachhuber committed:
Fall back to 'Untitled' when no title is present

westonruter committed:
Use currying function for de-duplicating toggling handlers

westonruter commented:
Found a styling issue with floats when the avatar alone is displayed:

![image](https://user-images.githubusercontent.com/134745/28740883-f6bfeae2-73bf-11e7-8e5c-950a3157a3ab.png)

danielbachhuber commented:
> Found a styling issue with floats when the avatar alone is displayed:

Does this need to be handled in this PR?

westonruter commented:
> Does this need to be handled in this PR?

Sure. Why not?

Also, one other big piece needed in this PR is the server-side component for rendering the comments when the block is displayed on the frontend. See #2014 which just refactored where the PHP logic is located.

westonruter commented:
Rebased with `master`. Rewrote 275b7d93 as aab2f81b.

westonruter committed:
Eliminate latency when decrementing comment count

Copied from Latest Posts block via https://github.com/WordPress/gutenberg/pull/1992/commits/3d9b34a7609f4124fb0438f052ba8e82eebbdc6a

westonruter committed:
Fix avatar floating and improve reliability of obtaining avatar size

westonruter committed:
Add initial server-side rendering of Latest Comments block

westonruter commented:
In the server-side block rendering, I'm confused as to why the `$attributes` passed to `gutenberg_render_block_core_latest_comments()` seem to ignore the attributes as provided in the content.

westonruter committed:
Add author name to comment list

westonruter commented:
@melchoyce here's how the block is now rendering in the editor:

![image](https://user-images.githubusercontent.com/134745/28853337-bd470b88-76e4-11e7-88ad-bbfab4c94ffa.png)

And how it renders on the frontend:

![image](https://user-images.githubusercontent.com/134745/28853366-e9ec7376-76e4-11e7-98f4-812ee2eb43cb.png)

Last major piece remaining is how to handle comment excerpts in the editor.

westonruter committed:
Fix tests for latest-comments attributes

melchoyce commented:
@westonruter Cool 👍 I have some visual tweaks I still want to make, but those can happen in a followup PR. 

Do you know if there's anything we can do to have the front-end view of the comments block inherit the theme's comment styles, or are we out of luck there?

westonruter commented:
@melchoyce:
> Do you know if there's anything we can do to have the front-end view of the comments block inherit the theme's comment styles, or are we out of luck there?

Yeah, I think it is possible. We'd essentially need to use the `wp_list_comments()` function (template tag) instead of manually crafting our own list. Or at least, we'd need to craft the list to use the same markup and class names (e.g. `.comment-body` and `.comment-author`) that the comment walker outputs, since actually `wp_list_comments()` only lists comments for the _current_ queried post and not all posts across the site. 

So that brings up another consideration. There's really two separate comment blocks. One block for the latest comments across the site (as built in this PR) and another block for the latest comments _for this current post_. This latter one I could see useful to show at the top of a longform post in a call-to-action to comment, or even in template building such a block could be used instead of `comments.php` entirely. This latter one would then also need to take into account threading and the integration of the comment form.

aduth commented on blocks/library/latest-comments/index.js:
```diff
@@ -0,0 +1,227 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from 'element';
```
Note with #2172, these need to be updated to prefix the dependencies with `@wordpress/`. You will need to perform a rebase against the latest version of master and apply your changes:

```
git fetch origin
git rebase origin/master
```

westonruter commented:
Rebased and added dependency prefixes. Rewrote e34aeeb as e0e9088.

westonruter commented:
Working on the PHP rendering of this block has confirmed for me the painfulness of having to implement two separate rendering systems, in PHP and JS: the former using normal/ubiquitous WordPress template tags and PHP API function calls, and the latter in React with calls to the REST API. I would not be able to suggest developers look at this block for a recommended streamlined way to write a dynamic block, given the violation of DRY. Maintaining two separate rendering systems with an aim to have the editor be a faithful preview of the frontend will be tedious.

mtias commented:
Are we good with merging this as a first pass?

> I would not be able to suggest developers look at this block for a recommended streamlined way to write a dynamic block, given the violation of DRY.

This is only repeating yourself if the editor is not offering editing capabilities for the block, in which case, by all means, people are free to server render it. If you have to write a native mobile block you will need to recreate the rendering as well, that's just the nature of optimizing for each experience.

In the case of widgets, is there a reason why the already built rendering can't be used on the back-end? That means only writing the editor implementation.

westonruter commented:
@mtias:

> Are we good with merging this as a first pass?

The only thing not implemented is truncating comment excerpts in the editor, and outputting the comment HTML in a safe way.

> This is only repeating yourself if the editor is not offering editing capabilities for the block, in which case, by all means, people are free to server render it. If you have to write a native mobile block you will need to recreate the rendering as well, that's just the nature of optimizing for each experience.

I guess it depends on what you mean by “editing capabilities”. Like is toggling whether or not the date is rendered such an edit?

> In the case of widgets, is there a reason why the already built rendering can't be used on the back-end? That means only writing the editor implementation.

This is exactly what I've been suggesting, I think. Using the built-in server-side PHP templating inside the editor block—the same logic that is used to render the dynamic block on the frontend—is what I think we should do in order to eliminate writing two separate renderers. In this way, the a dynamic widget block in the editor would be very similar to what was implemented for shortcode previews in the Shortcake plugin. It would also be very similar to how changes to server-rendered partials are previewed in the Customizer via selective refresh.

mtias commented:
> Using the built-in server-side PHP templating inside the editor block—the same logic that is used to render the dynamic block on the frontend—is what I think we should do in order to eliminate writing two separate renderers.

This is not what I mean. I mean using the already built server-side implementation as the front-end rendering instead of redoing all that again, and only creating the editor one from scratch. Server rendering blocks in the editor context doesn't seem a good direction, only a shortcut that would worsen the editing experience in favor of the development experience of only writing once.

westonruter commented:
@mtias: For the Recent Comments widget in core, its [server-side rendering logic](https://github.com/WordPress/wordpress-develop/blob/9ab26b8800e403c209599b7129fa32c9b7f1ac2a/src/wp-includes/widgets/class-wp-widget-recent-comments.php#L99-L121) doesn't itself specifically have filters to modify its markup. But another “Comments for this Post” block (as [described above](https://github.com/WordPress/gutenberg/pull/1931#issuecomment-319753975)) would indeed have filters since it would use `wp_list_comments()` which themes and plugins can modify the output of via the `wp_list_comments_args` filter (even supplying a completely custom `Walker_Comment`), besides all of the filters that the comment _form_ itself supports in `comment_form()`.

So for the Latest Comments widget block, a client-side React implementation of the UI _can_ indeed be implemented to mirror whatever rendering logic is developed on the server, but I fear this is not going to be scalable or possible in many cases. Mirroring implementations is re-work (not DRY). But if a theme could override a server-side `renderer` of a Latest Comments block to do something different, and in that case would it be expected that a theme must also create a custom block with its own client-side renderer as well?

If not, and in the case of a “Comments for this Post” block, the source of truth for the block's rendering is the server-side logic. There would not be a way to get a true preview of what the block will render without asking the server to render it. For static blocks, the source of truth _is_ the client-side renderer, so they are kept DRY. For dynamic blocks, this isn't the case. So I don't see server-side rendering of blocks as being a shortcut. It's about not creating more work than is required. It's about honoring the source of truth for what a dynamic block renders, and it's about giving a faithful preview of what the block actually will render. 

If many of a dynamic block's settings are purely stylistic, there needn't be a worsening of the editing experience as the client and server in that case could just toggle classes on the wrapper. Otherwise, if an attribute change does require a re-rendering then this is not really a fundamental change from what we have with a client-side rendering because in the case of client-side rendering there is still often a need to refetch data from the API, and so there will often be latency in that case as well.

I think using the server-side rendering of blocks in the editor will be key for developer adoption. Nobody wants to do work twice. It's also important for user adoption, because users want to preview what they're actually going to see rendered on the frontend.

danielbachhuber commented:
> I think using the server-side rendering of blocks in the editor will be key for developer adoption. Nobody wants to do work twice.

Just to throw an idea in because I haven't seen it discussed: what about a common templating implementation (e.g. Mustache) that can be used both server-side and client-side?

aduth commented:
Maybe of interest for consolidating/automating server-registered block attributes: #2529

westonruter commented:
@danielbachhuber The adoption of templating languages in Core has historically always been shied away from (e.g. Smarty, Twig, etc), the exception being Underscore templates in JS. WordPress uses “pure” PHP templates, for better or worse, and this is why JSX has been viewed so favorably so as to not be using any templating language. Adding a language-agnostic templating language would make the development experience worse on both the client and the server (lack of logic, lack of virtual DOM). I want to open a `try` PR fork of this branch that uses the Shortcake approach with DRY server-side rendering of the the dynamic block.

danielbachhuber commented:
Closing this PR for now. Feel free to re-open when you're ready to proceed on it.
